### PR TITLE
Pal 226 completable future feign bug

### PR DESCRIPTION
**Pull request by dev930018**

### CODE REVIEW

**Review by Leg60 — COMMENTED** [review]
Integration test now passes. Happy that a user is now returned.

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/service/PolicyService.java [hunk 1]
```diff
@@ -45,7 +43,7 @@ public PolicyService(final PolicyClient policyClient, final Executor executor) {
         CompletionStage<MultiPolicy> policy;
 
         try {
-            policy = this.client.getPolicy(policyRequest);
+            policy = CompletableFuture.supplyAsync(() -> this.client.getPolicy(policyRequest));
```
Comment (nw1984) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/service/PolicyService.java]:
```suggestion
            policy = CompletableFuture.supplyAsync(() -> client.getPolicy(policyRequest));
```

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/service/ResourceService.java [hunk 2]
```diff
@@ -44,7 +44,7 @@ public ResourceService(final ResourceClient resourceClient, final Executor execu
 
         CompletionStage<Map<LeafResource, ConnectionDetail>> resources;
         try {
-            resources = this.client.getResourcesById(resource);
+            resources = CompletableFuture.supplyAsync(() -> this.client.getResourcesById(resource));
```
Comment (nw1984) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/service/ResourceService.java]:
```suggestion
            resources = CompletableFuture.supplyAsync(() -> client.getResourcesById(resource));
```

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/service/UserService.java [hunk 3]
```diff
@@ -42,7 +42,7 @@ public UserService(final UserClient userClient, final Executor executor) {
 
         CompletionStage<User> user;
         try {
-            user = this.userClient.getUser(request);
+            user = CompletableFuture.supplyAsync(() -> this.userClient.getUser(request));
```
Comment (nw1984) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/service/UserService.java]:
```suggestion
            user = CompletableFuture.supplyAsync(() -> userClient.getUser(request));
```

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/ResourceClient.java [hunk 4]
```diff
@@ -27,22 +27,21 @@
 import uk.gov.gchq.palisade.service.palisade.request.GetResourcesByTypeRequest;
 
 import java.util.Map;
-import java.util.concurrent.CompletableFuture;
 
 @FeignClient(value = "resource-service", url = "${web.client.resource-service}")
 public interface ResourceClient {
```
Comment (nw1984) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/ResourceClient.java]: I question the merits of returning a synchronous function for all of the ResourceService - some of these resources could take a long time to complete.

Spring guide here on implementing async methods:
https://spring.io/guides/gs/async-method/

Any test harness that uses the async method will have to do an asyncDispatch() to ensure that the async call is completed.

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/UserClient.java [hunk 5]
```diff
@@ -21,12 +21,10 @@
...
 import uk.gov.gchq.palisade.service.palisade.request.GetUserRequest;
 
-import java.util.concurrent.CompletableFuture;
-
 @FeignClient(value = "user-service", url = "${web.client.user-service}")
 public interface UserClient {
 
     @PostMapping(value = "/getUser", consumes = "application/json", produces = "application/json")
-    CompletableFuture<User> getUser(final GetUserRequest request);
+    User getUser(final GetUserRequest request);
```
Comment (nw1984) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/web/UserClient.java:28]: ditto above.

**Review by nw1984 — CHANGES_REQUESTED** [review]

**Review by m09526 — APPROVED** [review]

**Review by dev-958 — APPROVED** [review]

**Pull request merged**